### PR TITLE
Total Token Transfers on Optimism

### PR DESCRIPTION
**Pull request by MSilb7**
Brief comments on the purpose of your changes:

Spell to combine ETH, ERC20, NFT transfers in to one table, with some extra fields to describe the transaction type (i.e. internal transfer, transfer transaction, mint, burn).

Still thinking about better column names than 'transfer_style', 'transfer_tx_type', etc

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by MSilb7** [comment]
Should be good now, also tagging @hildobby and @chuxinh since they worked on some of the underlying transfer queries. Does this make sense, or are parts dumb?

+ Any help on column naming ideas would be nice.

File: models/transfers/optimism/transfers_optimism_tokens.sql [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        token_type,
+        token_id,
+        transfer_type,
+        -- mint, burn, or transfer
+        CASE
+                WHEN transfer_from_address = '0x0000000000000000000000000000000000000000' THEN 'mint'
+                WHEN transfer_to_address = '0x0000000000000000000000000000000000000000' THEN 'burn'
+                ELSE 'transfer'
+                END 
+        AS transfer_style,
```
Comment (chuxinh) [inline models/transfers/optimism/transfers_optimism_tokens.sql:28]: wonder if `transfer_action` here is better?

File: models/transfers/optimism/transfers_optimism_tokens.sql [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                        THEN 'transfer transaction'
+                WHEN t.gas_used = 21000 AND substring(t.data, 1, 10) = '0x'
+                        THEN 'transfer transaction' -- we specify gas_used because some arb bots will send null data as well.
+                WHEN substring(t.data, 1, 10) = '0xd0e30db0'
+                        THEN 'eth wrap'
+                WHEN substring(t.data, 1, 10) = '0x2e1a7d4d'
+                        THEN 'eth unwrap'
+                ELSE 'internal transaction'
+                END
+        AS transfer_tx_type,
```
Comment (chuxinh) [inline models/transfers/optimism/transfers_optimism_tokens.sql:45]: maybe just `tx_type`

File: models/transfers/optimism/transfers_optimism_tokens.sql (hunk 1 above)
Comment (MSilb7) [reply models/transfers/optimism/transfers_optimism_tokens.sql:28]: i like it

File: models/transfers/optimism/transfers_optimism_tokens.sql [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        
+        r.value,
+        r.evt_block_time AS tx_block_time,
+        r.evt_block_number AS tx_block_number,
+        r.evt_tx_hash AS tx_hash,
+
+        r.evt_index,
+        '' AS trace_address,
+
+        r.evt_tx_hash || '-' || CAST(evt_index AS VARCHAR(100))  as unique_transfer_id
```
Comment (Hosuke) [inline models/transfers/optimism/transfers_optimism_tokens.sql:82]: Hi @MSilb7, the `transfers_optimism_tokens` looks lgtm. though.
We may define the new `unique_key` ( like [evt_tx_hash, evt_index, trace_address] ) for incremental merge in this model, as `unique_transfer_id` may be a deprecated feature.
What do you think?

File: models/transfers/optimism/transfers_optimism_tokens.sql [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        -----------
+        SELECT
+
+        r.from AS transfer_from_address,
+        r.to AS transfer_to_address,
+        r.contract_address AS contract_address,
+        'erc20' AS token_standard,
+        'fungible' AS token_type,
+        '' AS token_id, -- used by NFTs
+        '' AS transfer_type, -- used by NFTs
```
Comment (Hosuke) [inline models/transfers/optimism/transfers_optimism_tokens.sql:71]: Also this part for `transfer_type` shoule be 'single' or 'batch' ?
If none of these we may need to add new accepted values in schema.

File: models/transfers/optimism/transfers_optimism_tokens.sql (hunk 3 above)
Comment (MSilb7) [reply models/transfers/optimism/transfers_optimism_tokens.sql:82]: you mean 'unique_transfer_id' may be deprecated in other tables? I'm good removing it.

File: models/transfers/optimism/transfers_optimism_tokens.sql (hunk 4 above)
Comment (MSilb7) [reply models/transfers/optimism/transfers_optimism_tokens.sql:71]: I'll remove the test. This should be handled for in the NFT tables anyway.

**Comment by Hosuke** [comment]
I added uniqueness test for token transfers in schema in last commit.

**Comment by Hosuke** [comment]
Hi @MSilb7, there are dupes produced in the current model, and we need fix this.
https://github.com/duneanalytics/spellbook/actions/runs/4300948232/jobs/7497642147#step:10:43
Detailed duped records:
https://dune.com/queries/2099337?d=1

**Comment by MSilb7** [comment]
> Hi @MSilb7, there are dupes produced in the current model, and we need fix this. https://github.com/duneanalytics/spellbook/actions/runs/4300948232/jobs/7497642147#step:10:43 Detailed duped records: https://dune.com/queries/2099337?d=1

1155s smh. We can add token_id to the uniqueness check I guess

**Comment by jeff-dude** [comment]
it appears the merge is failing, as duplicates explode after the incremental run. if stuck on why, i'd advise:
- compile query
- write query on dune which has multiple CTEs (one as source with incremental filter applied, one as target for full query, one to join the two on unique keys)
- find out why join is duplicating the rows when it shouldn't (i.e. the rows which are re-inserting, rather than updating)

**Pull request closed without merging**